File: app/src/main/java/niuedu/com/treeviewtest/ExampleListTreeAdapter.java
package niuedu.com.treeviewtest;

import android.graphics.Bitmap;
import android.support.v7.widget.PopupMenu;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import com.niuedu.ListTree;
import com.niuedu.ListTreeAdapter;

/**
 * 为RecyclerView提供数据
 */
public class ExampleListTreeAdapter extends
        ListTreeAdapter<ExampleListTreeAdapter.BaseViewHolder> {

    //行上弹出菜单的侦听器
    private PopupMenu.OnMenuItemClickListener itemMenuClickListener;
    //记录弹出菜单是在哪个行上出现的
    private ListTree.TreeNode currentNode;

    //保存子行信息的类
    public static class ContactInfo{
        //头像,用于设置给ImageView。
        private Bitmap bitmap;
        //标题
        private String title;
        //描述
        private String detail;

        public ContactInfo(Bitmap bitmap, String title, String detail) {
            this.bitmap = bitmap;
            this.title = title;
            this.detail = detail;
        }

        public Bitmap getBitmap() {
            return bitmap;
        }

        public String getTitle() {
            return title;
        }
        public void setTitle(String title) {
            this.title = title;
        }

        public String getDetail() {
            return detail;
        }

    }

    //构造方法
    public ExampleListTreeAdapter(ListTree tree,
                                  PopupMenu.OnMenuItemClickListener listener){
        super(tree);
        this.itemMenuClickListener=listener;
    }

    public ListTree.TreeNode getCurrentNode() {
        return currentNode;
    }

    @Override
    protected BaseViewHolder onCreateNodeView(ViewGroup parent, int viewType){
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        //创建不同的行View
        if(viewType==R.layout.contacts_group_item){
            //注意！此处有一个不同！最后一个参数必须传true！
            View view = inflater.inflate(viewType,parent,true);
            //用不同的ViewHolder包装
            return new GroupViewHolder(view);
        }else if(viewType == R.layout.contacts_contact_item){
            //注意！此处有一个不同！最后一个参数必须传true！
            View view = inflater.inflate(viewType,parent,true);
            //用不同的ViewHolder包装
            return  new ContactViewHolder(view);
        }else{
            return null;
        }
    }

    @Override
    protected void onBindNodeViewHolder(BaseViewHolder holder, int position) {
        View view = holder.itemView;
        //get node at the position
        ListTree.TreeNode node = tree.getNodeByPlaneIndex(position);

        if(node.getLayoutResId() == R.layout.contacts_group_item){
            //group node
            String title = (String)node.getData();

            GroupViewHolder gvh= (GroupViewHolder) holder;
            gvh.textViewTitle.setText(title);
            gvh.textViewCount.setText("0/"+node.getChildrenCount());
            gvh.aSwitch.setChecked(node.isChecked());
        }else if(node.getLayoutResId() == R.layout.contacts_contact_item){
            //child node
            ContactInfo info = (ContactInfo) node.getData();

            ContactViewHolder cvh= (ContactViewHolder) holder;
            cvh.imageViewHead.setImageBitmap(info.getBitmap());
            cvh.textViewTitle.setText(info.getTitle());
            cvh.textViewDetail.setText(info.getDetail());
            cvh.aSwitch.setChecked(node.isChecked());
        }
    }

    //组行和联系人行的Holder基类
    class BaseViewHolder extends ListTreeAdapter.ListTreeViewHolder{
        public BaseViewHolder(View itemView) {
            super(itemView);
        }
    }

    //将ViewHolder声明为Adapter的内部类，反正外面也用不到
    class GroupViewHolder extends BaseViewHolder {

        TextView textViewTitle;
        TextView textViewCount;
        Switch aSwitch;
        TextView textViewMenu;

        public GroupViewHolder(View itemView) {
            super(itemView);

            textViewTitle = itemView.findViewById(R.id.textViewTitle);
            textViewCount = itemView.findViewById(R.id.textViewCount);
            aSwitch = itemView.findViewById(R.id.switchChecked);
            textViewMenu = itemView.findViewById(R.id.textViewMenu);

            //应响应点击事件而不是CheckedChange事件，因为那样会引起事件的递归触发
            aSwitch.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    int planeIndex = getAdapterPosition();
                    ListTree.TreeNode node = tree.getNodeByPlaneIndex(planeIndex);
                    node.setChecked(!node.isChecked());
                    //改变所有的子孙们的状态
                    int count =tree.setDescendantChecked(planeIndex,node.isChecked());
                    notifyItemRangeChanged(planeIndex,count+1);
                }
            });

            //点了PopMenu控件，弹出PopMenu
            textViewMenu.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int nodePlaneIndex = getAdapterPosition();
                    ListTree.TreeNode node = tree.getNodeByPlaneIndex(nodePlaneIndex);
                    currentNode=node;
                    PopupMenu popup = new PopupMenu(v.getContext(), v);
                    popup.setOnMenuItemClickListener(itemMenuClickListener);
                    MenuInflater inflater = popup.getMenuInflater();
                    inflater.inflate(R.menu.menu_item, popup.getMenu());
                    popup.show();
                }
            });
        }
    }

    class ContactViewHolder extends BaseViewHolder{
        ImageView imageViewHead;
        TextView textViewTitle;
        TextView textViewDetail;
        Switch aSwitch;

        public ContactViewHolder(View itemView) {
            super(itemView);

            imageViewHead = itemView.findViewById(R.id.imageViewHead);
            textViewTitle = itemView.findViewById(R.id.textViewTitle);
            textViewDetail = itemView.findViewById(R.id.textViewDetail);
            aSwitch = itemView.findViewById(R.id.switchChecked);

            //应响应点击事件而不是CheckedChange事件，因为那样会引起事件的递归触发
            aSwitch.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    int nodePlaneIndex = getAdapterPosition();
                    ListTree.TreeNode node = tree.getNodeByPlaneIndex(nodePlaneIndex);
                    node.setChecked(!node.isChecked());
                    //改变所有的子孙们的状态
                    int count =tree.setDescendantChecked(nodePlaneIndex,node.isChecked());
                    notifyItemRangeChanged(nodePlaneIndex,count+1);
                }
            });
        }
    }
}

File: app4cotlin/src/main/java/qst/com/app4cotlin/ExampleListTreeAdapter.kt
package qst.com.app4cotlin

import android.graphics.Bitmap
import android.support.v7.widget.PopupMenu
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.Switch
import android.widget.TextView
import com.niuedu.ListTree
import com.niuedu.ListTreeAdapter

class ExampleListTreeAdapter(tree: ListTree, listener : PopupMenu.OnMenuItemClickListener) :
        ListTreeAdapter<ExampleListTreeAdapter.BaseViewHolder>(tree){

    //行上弹出菜单的侦听器
    private val itemMenuClickListener : PopupMenu.OnMenuItemClickListener

    //记录弹出菜单是在哪个行上出现的
    var currentNode: ListTree.TreeNode? = null

    //保存子行信息的类
    class ContactInfo(
            val bitmap: Bitmap, //头像,用于设置给ImageView
            var title: String, //标题
            var detail: String //描述
    )

    init{
        itemMenuClickListener = listener
    }

    override fun onCreateNodeView(parent: ViewGroup?, viewType: Int): BaseViewHolder? {
        val inflater = LayoutInflater.from(parent!!.getContext())

        //创建不同的行View
        if (viewType == R.layout.contacts_group_item) {
            //注意！此处有一个不同！最后一个参数必须传true！
            val view = inflater.inflate(viewType, parent, true)
            //用不同的ViewHolder包装
            return GroupViewHolder(view)
        } else if (viewType == R.layout.contacts_contact_item) {
            //注意！此处有一个不同！最后一个参数必须传true！
            val view = inflater.inflate(viewType, parent, true)
            //用不同的ViewHolder包装
            return ContactViewHolder(view)
        } else {
            return null
        }
    }

    override fun onBindNodeViewHolder(viewHoler: BaseViewHolder?, position: Int) {
        //get node at the position
        val node = tree.getNodeByPlaneIndex(position)

        if (node.layoutResId == R.layout.contacts_group_item) {
            //group node
            val title = node.data as String

            val gvh = viewHoler as GroupViewHolder
            gvh.textViewTitle.text = title
            gvh.textViewCount.text = "0/" + node.childrenCount
            gvh.aSwitch.isChecked = node.isChecked
        } else if (node.layoutResId == R.layout.contacts_contact_item) {
            //child node
            val info = node.data as ContactInfo

            val cvh = viewHoler as ContactViewHolder
            cvh.imageViewHead.setImageBitmap(info.bitmap)
            cvh.textViewTitle.text = info.title
            cvh.textViewDetail.text = info.detail
            cvh.aSwitch.isChecked = node.isChecked
        }
    }

    //组行和联系人行的Holder基类
    open inner class BaseViewHolder(itemView: View) : ListTreeViewHolder(itemView)

    //将ViewHolder声明为Adapter的内部类，反正外面也用不到
    internal inner class GroupViewHolder(itemView: View) : BaseViewHolder(itemView) {

        var textViewTitle: TextView = itemView.findViewById(R.id.textViewTitle)
        var textViewCount: TextView = itemView.findViewById(R.id.textViewCount)
        var aSwitch: Switch = itemView.findViewById(R.id.switchChecked)
        var textViewMenu: TextView = itemView.findViewById(R.id.textViewMenu)

        init {

            //应响应点击事件而不是CheckedChange事件，因为那样会引起事件的递归触发
            aSwitch.setOnClickListener {
                val planeIndex = adapterPosition
                val node = tree.getNodeByPlaneIndex(planeIndex)
                node.isChecked = !node.isChecked
                //改变所有的子孙们的状态
                val count = tree.setDescendantChecked(planeIndex, node.isChecked)
                notifyItemRangeChanged(planeIndex, count + 1)
            }

            //点了PopMenu控件，弹出PopMenu
            textViewMenu.setOnClickListener { v ->
                val nodePlaneIndex = adapterPosition
                val node = tree.getNodeByPlaneIndex(nodePlaneIndex)
                currentNode = node
                val popup = PopupMenu(v.context, v)
                popup.setOnMenuItemClickListener(itemMenuClickListener)
                val inflater = popup.menuInflater
                inflater.inflate(R.menu.menu_item, popup.menu)
                popup.show()
            }
        }
    }

    internal inner class ContactViewHolder(itemView: View) : BaseViewHolder(itemView) {
        var imageViewHead: ImageView = itemView.findViewById(R.id.imageViewHead)
        var textViewTitle: TextView = itemView.findViewById(R.id.textViewTitle)
        var textViewDetail: TextView = itemView.findViewById(R.id.textViewDetail)
        var aSwitch: Switch = itemView.findViewById(R.id.switchChecked)

        init {

            //应响应点击事件而不是CheckedChange事件，因为那样会引起事件的递归触发
            aSwitch.setOnClickListener {
                val nodePlaneIndex = adapterPosition
                val node = tree.getNodeByPlaneIndex(nodePlaneIndex)
                node.isChecked = !node.isChecked
                //改变所有的子孙们的状态
                val count = tree.setDescendantChecked(nodePlaneIndex, node.isChecked)
                notifyItemRangeChanged(nodePlaneIndex, count + 1)
            }
        }
    }

}